db/repair_bucket.go

```diff
@@ -0,0 +1,228 @@
+package db
+
+import (
+	"encoding/json"
+
+	"io/ioutil"
+
+	"fmt"
+
+	"github.com/couchbase/go-couchbase"
+	"github.com/couchbase/sync_gateway/base"
+)
+
+// Enum for the different repair jobs (eg, repairing rev tree cycles)
+type RepairJobType string
+
+const (
+	RepairRevTreeCycles = RepairJobType("RepairRevTreeCycles")
+)
+
+// Params suitable for external (eg, HTTP) invocations to describe a RepairBucket operation
+type RepairBucketParams struct {
+	DryRun     bool              `json:"dry_run"`
+	RepairJobs []RepairJobParams `json:"repair_jobs"`
+}
+
+// Params suitable for external (eg, HTTP) invocations to describe a specific RepairJob operation
+type RepairJobParams struct {
+	RepairJobType   RepairJobType          `json:"repair_job_type"`
+	RepairJobParams map[string]interface{} `json:"repair_job_params"`
+}
+
+// Given a Couchbase Bucket doc, transform the doc in some way to produce a new doc.
+// Also return a boolean to indicate whether a transformation took place, or any errors occurred.
+type DocTransformer func(doc *document) (transformedDoc *document, transformed bool, err error)
+
+// A RepairBucket struct is the main API entrypoint to call for repairing documents in buckets
+type RepairBucket struct {
+	DryRun                  bool // If true, will only output what changes it *would* have made, but not make any changes
+	WriteRepairedDocsToDisk bool // If true, write the before and after doc to temp files
+	Bucket                  base.Bucket
+	RepairJobs              []DocTransformer
+}
+
+func NewRepairBucket(bucket base.Bucket) *RepairBucket {
+	return &RepairBucket{
+		Bucket: bucket,
+	}
+}
+
+func (r *RepairBucket) SetDryRun(dryRun bool) *RepairBucket {
+	r.DryRun = dryRun
+	return r
+}
+
+func (r *RepairBucket) AddRepairJob(repairJob DocTransformer) *RepairBucket {
+	r.RepairJobs = append(r.RepairJobs, repairJob)
+	return r
+}
+
+func (r *RepairBucket) InitFrom(params RepairBucketParams) *RepairBucket {
+
+	r.SetDryRun(params.DryRun)
+	r.WriteRepairedDocsToDisk = true
+	for _, repairJobParams := range params.RepairJobs {
+		switch repairJobParams.RepairJobType {
+		case RepairRevTreeCycles:
+			r.AddRepairJob(RepairJobRevTreeCycles)
+		}
+	}
+
+	return r
+}
+
+func (r RepairBucket) RepairBucket() (repairedDocs []*document, err error) {
```

Comment:
If the set of docs being repaired is large, do we have memory concerns associated with the repairedDocs array?

Reply:
Changed to return doc ids